cli/npm/permission_checker.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::io::ErrorKind;
+use std::path::Path;
+use std::path::PathBuf;
+
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use deno_core::parking_lot::Mutex;
+use deno_runtime::deno_node::NodePermissions;
+use sys_traits::FsCanonicalize;
+
+use crate::sys::CliSys;
+
+#[derive(Debug)]
+pub enum NpmRegistryReadPermissionCheckerMode {
+  Byonm,
+  Global(PathBuf),
+  Local(PathBuf),
+}
+
+#[derive(Debug)]
+pub struct NpmRegistryReadPermissionChecker {
+  sys: CliSys,
+  cache: Mutex<HashMap<PathBuf, PathBuf>>,
+  mode: NpmRegistryReadPermissionCheckerMode,
+}
+
+impl NpmRegistryReadPermissionChecker {
+  pub fn new(sys: CliSys, mode: NpmRegistryReadPermissionCheckerMode) -> Self {
+    Self {
+      sys,
+      cache: Default::default(),
+      mode,
+    }
+  }
+
+  #[must_use = "the resolved return value to mitigate time-of-check to time-of-use issues"]
+  pub fn ensure_read_permission<'a>(
+    &self,
+    permissions: &mut dyn NodePermissions,
+    path: &'a Path,
+  ) -> Result<Cow<'a, Path>, AnyError> {
+    if permissions.query_read_all() {
+      return Ok(Cow::Borrowed(path)); // skip permissions checks below
+    }
+
+    match &self.mode {
+      NpmRegistryReadPermissionCheckerMode::Byonm => {
```

Comment:
Change here was to also have the byonm check here, but other than that the code is the same.